external-crates/move/crates/move-disassembler/src/disassembler.rs

```diff
@@ -197,48 +227,83 @@ impl<'a> Disassembler<'a> {
         self.coverage_map = Some(coverage_map);
     }
 
+    /// Disassemble the module and return the disassembled string.
     pub fn disassemble(&self) -> Result<String> {
         let mut buffer = String::new();
+        let bcode_map_gen = false;
         if let Some(budget) = self.options.max_output_size {
-            self.print_module(&mut BoundedBuffer {
-                buf: &mut buffer,
-                budget,
-            })
+            self.print_module(
+                &mut BoundedBuffer {
+                    buf: &mut buffer,
+                    budget,
+                },
+                bcode_map_gen,
+            )
             .map_err(|e| anyhow::anyhow!("{e}: Module exceeded max allowed disassembly size"))?;
         } else {
-            self.print_module(&mut buffer)?;
+            self.print_module(&mut buffer, bcode_map_gen)?;
         };
         Ok(buffer)
     }
+
+    /// Disassemble the module and return the disassembled string,
+    /// but also a source map for the disassembled bytecode.
+    pub fn disassemble_with_source_map(&self) -> Result<(String, SourceMap)> {
+        let mut buffer = String::new();
+        let bcode_map_gen = true;
+        let mut bcode_map = if let Some(budget) = self.options.max_output_size {
+            self.print_module(
+                &mut BoundedBuffer {
+                    buf: &mut buffer,
+                    budget,
+                },
+                bcode_map_gen,
+            )
+            .map_err(|e| anyhow::anyhow!("{e}: Module exceeded max allowed disassembly size"))?
+        } else {
+            self.print_module(&mut buffer, bcode_map_gen)?
+        };
+        let file_hash = FileHash::new(&buffer);
+        bcode_map.replace_file_hashes(file_hash);
+        Ok((buffer, bcode_map))
+    }
 }
 
 // Note on naming:
 // * disassemble_* and print_* functions are functions that output to the buffer
 // * format_* functions return a string that can be used in the buffer
 impl<'a> Disassembler<'a> {
-    fn print_module(&self, buffer: &mut impl Write) -> Result<()> {
+    fn print_module(
+        &self,
+        buffer: &mut (impl Write + ByteLength),
+        bcode_map_gen: bool,
+    ) -> Result<SourceMap> {
         // NB: The order in which these are called is important as each function is effectful.
-        self.print_header(buffer)?;
+        let mut bcode_map = self.print_header(buffer)?;
```

Comment:
See comment below but I think turning this into an option may make things even cleaner.